notifier/notifier.go

```diff
@@ -221,43 +220,118 @@ func (n *Notifier) collectRogueCallbacks() {
 	}
 }
 
-// Notify posts callback info to j.CallbackURL
-func (n *Notifier) Notify(j *job.Job) error {
+// MonitorDeliveries starts as many go routines as limit instructs in order to
+// poll/monitor the delivery reports channel.
+func (n *Notifier) MonitorDeliveries(ctx context.Context, limit int, dwg *sync.WaitGroup) {
+	dwg.Add(limit)
+
+	for i := 0; i < limit; i++ {
+		go n.monitorDeliveriesRoutine(ctx, dwg)
+	}
+}
+
+// monitorDeliveriesRoutine watches events from the delivery reports channel and
+// also checks whether the context has been cancelled or not.
+func (n *Notifier) monitorDeliveriesRoutine(ctx context.Context, dwg *sync.WaitGroup) {
+	defer dwg.Done()
+
+	for {
+		select {
+		case <-ctx.Done():
+			n.Log.Println("Context has been cancelled")
+			return
+		case httpInfo := <-n.backends["http"].DeliveryReports():
+			n.handleCallbackInfo(httpInfo)
+			continue
+		case kafkaInfo := <-n.backends["kafka"].DeliveryReports():
+			n.handleCallbackInfo(kafkaInfo)
+			continue
+		}
+	}
+}
+
+// handleCallbackInfo handles a callback info object.
+// If the callback is successful we increment stats counters and remove the job from storage.
+// For kafka if the callback is not successful we immediately mark the callback as failed since
+// kafka has already tried to retransmit a message for an amount of time. For http if the
+// callback is not successful we have already handled this case in Notify() method with retryOrFail().
+// Also not that when on http we call QueueJobForDeletion which schedules file deletion associated to our job,
+// because we are sure the other end has downloaded the file. But kakfa only ensures that a message has been
+// persisted, but most likely consumer has not recieved the message yet alone the file.
+func (n *Notifier) handleCallbackInfo(cbInfo job.CallbackInfo) {
+	j, err := n.Storage.GetJob(cbInfo.JobID)
+	if err != nil {
+		n.Log.Printf("Error: Could not get job %s. Operation returned error: %s", cbInfo.JobID, err)
```

Comment:
Generally, prefer to return errors instead of just logging them. Then, let them bubble up to the callers, where you will handle them.